### PR TITLE
feature: add rsync task

### PR DESCRIPTION
**Pull request by shaowenchen**
Signed-off-by: shaowenchen <mail@chenshaowen.com>

# Changes

add sync task

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Follows the [authoring recommendations](https://github.com/tektoncd/catalog/blob/main/recommendations.md)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality of task changed or new task added)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Complies with [Catalog Organization TEP][TEP], see [example]. **Note** [An issue has been filed to automate this validation][validation]
  - [x] File path follows  `<kind>/<name>/<version>/name.yaml`
  - [x] Has `README.md` at `<kind>/<name>/<version>/README.md`
  - [x] Has mandatory `metadata.labels` - `app.kubernetes.io/version` the same as the `<version>` of the resource
  - [x] Has mandatory `metadata.annotations` `tekton.dev/pipelines.minVersion`
  - [x] mandatory `spec.description` follows the convention

          ```

          spec:
            description: >-
              one line summary of the resource

              Paragraph(s) to describe the resource.
          ```

_See [the contribution guide](https://github.com/tektoncd/catalog/blob/master/CONTRIBUTING.md)
for more details._

---

[TEP]: https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md
[example]: https://github.com/tektoncd/catalog/tree/master/task/git-clone/0.1
[validation]:  https://github.com/tektoncd/catalog/issues/413

### CODE REVIEW

File: task/rsync/0.1/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
+ARG alpine_ver=3.13
+FROM alpine:${alpine_ver}.5
```
Comment (chmouel) [inline task/rsync/0.1/Dockerfile:2]: maybe you just want [3.13](https://hub.docker.com/layers/alpine/library/alpine/3.13/images/sha256-9663906b1c3bf891618ebcac857961531357525b25493ef717bca0f86f581ad6?context=explore) (or alpine_ver) to get all the minor release  and have ARG working

File: task/rsync/0.1/README.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
+# Rsync
+
+This task provides synchronization of files between container and remote host. It relies on rsync and sshpass commands to implement.
```
Comment (chmouel) [inline task/rsync/0.1/README.md:3]: it's maybe not the most secure way to do rsync, what about ssh key exchange instead?

a username/password leaking is worse than a key that can be confined in authorized_keys

File: task/rsync/0.1/README.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+            requests:
+              storage: 100M
+```
+
+
+- View remote host directory
+
+```bash
+[root@myk8s ~]# ls -al /root/
+
```
Comment (chmouel) [inline task/rsync/0.1/README.md:92]: I don't think we need to copy the whole outputs in there.

**Review by chmouel — CHANGES_REQUESTED** [review]
What about a test ? assuming you can spin up a ssh service in the test like the way we do for the jenkins task.

File: task/rsync/0.1/Dockerfile (hunk 1 above)
Comment (shaowenchen) [reply task/rsync/0.1/Dockerfile]: fixed.

File: task/rsync/0.1/README.md (hunk 2 above)
Comment (shaowenchen) [reply task/rsync/0.1/README.md:3]: I will remove username/password and replace it by authorized_keys.

File: task/rsync/0.1/README.md (hunk 3 above)
Comment (shaowenchen) [reply task/rsync/0.1/README.md:92]: Do you have any better suggestions? Here is just an example.

Comment (chmouel) [reply task/rsync/0.1/README.md:92]: well, you can just omit the output.

Comment (shaowenchen) [reply task/rsync/0.1/README.md]: OK

File: task/rsync/0.1/README.md (hunk 2 above)
Comment (shaowenchen) [reply task/rsync/0.1/README.md:3]: I have added ssh authentication.

**Comment by chmouel** [comment]
Thanks for making it more secure, I don't have time to do the full review now, but remove my hold so others can review into it

/hold cancel

**Comment by shaowenchen** [comment]
hi, @vdemeester  can you help to review this pr?

**Review by vdemeester — APPROVED** [review]

**Comment by vinamra28** [comment]
@shaowenchen can you please add yourself to the org so that the OWNER file can be made valid? Reference PR -> https://github.com/tektoncd/community/pull/166

**Comment by shaowenchen** [comment]
> @shaowenchen can you please add yourself to the org so that the OWNER file can be made valid? Reference PR -> [tektoncd/community#166](https://github.com/tektoncd/community/pull/166)

OK, I have submitted an application. After joining the organization, I will add the OWNER file to improve this pr.

File: task/rsync/0.1/rsync.yaml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      default: "./"
+    - name: remote-ip
+      type: string
+      default: "0.0.0.0"
+    - name: remote-username
+      type: string
+      default: "root"
+    - name: remote-password
+      type: string
+      default: "password"
```
Comment (vinamra28) [inline task/rsync/0.1/rsync.yaml:36]: shouldn't these be passed via secrets as passing the creds via params is the less secure way?

File: task/rsync/0.1/rsync.yaml [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    - name: remote-path
+      type: string
+      default: "~/"
+  steps:
+    - name: rsync
+      image: docker.io/shaowenchen/rsync-sshpass:v1
```
Comment (vinamra28) [inline task/rsync/0.1/rsync.yaml:42]: can we tag the image with digest as well?

File: task/rsync/0.1/rsync.yaml (hunk 4 above)
Comment (shaowenchen) [reply task/rsync/0.1/rsync.yaml:36]: How about using Kubernetes Secret to pass the account password here? I want to support both username/password and ssh auth.

File: task/rsync/0.1/rsync.yaml (hunk 5 above)
Comment (shaowenchen) [reply task/rsync/0.1/rsync.yaml]: fixed.

File: task/rsync/0.1/rsync.yaml (hunk 4 above)
Comment (vinamra28) [reply task/rsync/0.1/rsync.yaml:36]: yeah Kubernetes Secret will be good :+1:

Comment (shaowenchen) [reply task/rsync/0.1/rsync.yaml]: I test this with Kubernetes Secret, but I found the password printed out in the logs. I don't find a good path to hide it. So  I remove the basic auth by username/password.

Maybe, Tekton should provide a way to hide sensitive information.

Comment (shaowenchen) [reply task/rsync/0.1/rsync.yaml]: @vinamra28

**Comment by PuneetPunamiya** [comment]
test .*

File: task/rsync/0.1/rsync.yaml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: rsync
+  labels:
+    app.kubernetes.io/version: "0.1"
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.12.1"
+    tekton.dev/tags: rsync, files sync
+    tekton.dev/displayName: "rsync"
```
Comment (vinamra28) [inline task/rsync/0.1/rsync.yaml:11]: can you please add one more annotation `tekton.dev/categories` and if the image used is a multi-arch image then can you please add `tekton.dev/platforms` annotation?

File: task/rsync/0.1/rsync.yaml [hunk 7]
```diff
@@ -0,0 +1,55 @@
+---
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: rsync
+  labels:
+    app.kubernetes.io/version: "0.1"
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.12.1"
```
Comment (vinamra28) [inline task/rsync/0.1/rsync.yaml:9]: also since optional workspaces are being used over here can you please change the `pipelines.minVersion` to `0.17.x`

Comment (shaowenchen) [reply task/rsync/0.1/rsync.yaml]: fixed.

File: task/rsync/0.1/rsync.yaml (hunk 6 above)
Comment (shaowenchen) [reply task/rsync/0.1/rsync.yaml:12]: fixed.

**Review by vinamra28 — APPROVED** [review]
/meow

**Pull request merged**